File: design_patterns/C++/Creational/abstractFactory.cxx
#include <iostream>

class Door {
  public:
	virtual void getDescription() = 0;
};

class WoodenDoor: public Door {
  public:
	void getDescription() {
		std::cout << "I am a wooden door" << std::endl;
	}
};

class IronDoor: public Door {
  public:
	void getDescription() {
		std::cout << "I am an iron door" << std::endl;
	}
};

// fitting experts for each door type
class DoorFittingExpert {
  public:
	virtual void getDescription() = 0;
};

class Carpenter: public DoorFittingExpert {
  public:
	void getDescription() {
		std::cout << "I can only fit wooden doors" << std::endl;
	}
};

class Welder: public DoorFittingExpert {
  public:
	void getDescription() {
		std::cout << "I can only fit iron doors" << std::endl;
	}
};

// abstract factory
class DoorFactory {
  public:
	virtual Door* makeDoor() = 0;
	virtual DoorFittingExpert* makeFittingExpert() = 0;
};

// Wooden factory to return carpenter and wooden door
class WoodenDoorFactory: public DoorFactory {
  public:
	Door* makeDoor() {
		return new WoodenDoor();
	}

	DoorFittingExpert* makeFittingExpert() {
		return new Carpenter();
	}
};

// Iron door factory to get iron door and the relevant fitting expert
class IronDoorFactory: public DoorFactory {
  public:
	Door* makeDoor() {
		return new IronDoor();
	}

	DoorFittingExpert* makeFittingExpert() {
		return new Welder();
	}
};

int main() {
	auto woodenFactory = new WoodenDoorFactory();
	auto door1 = woodenFactory->makeDoor();
	auto expert1 = woodenFactory->makeFittingExpert();
	door1->getDescription();
	expert1->getDescription();
	// ===========================
	// I am a wooden door
	// I can only fit wooden doors
	// ===========================
	auto ironFactory = new IronDoorFactory();
	auto door2 = ironFactory->makeDoor();
	auto expert2 = ironFactory->makeFittingExpert();
	door2->getDescription();
	expert2->getDescription();
	// =========================
	// I am an iron door
	// I can only fit iron doors
	// =========================
	return 0;
}

File: design_patterns/Python/Creational/abstract_factory.py
class Door:

    def get_description(self):
        raise NotImplementedError


class WoodenDoor(Door):

    def get_description(self):
        print('I am a wooden door')


class IronDoor(Door):

    def get_description(self):
        print('I am a iron door')


class DoorFittingexpert:

    def get_description(self):
        raise NotImplementedError


class Welder(DoorFittingexpert):

    def get_description(self):
        print('I can only fit iron doors')


class Carpenter(DoorFittingexpert):

    def get_description(self):
        print('I can only fit wooden doors')


class DoorFactory:

    def make_door(self):
        raise NotImplementedError

    def make_fitting_expert(self):
        raise NotImplementedError


class WoodenDoorFactory(DoorFactory):

    def make_door(self):
        return WoodenDoor()

    def make_fitting_expert(self):
        return Carpenter()


class IronDoorFactory(DoorFactory):

    def make_door(self):
        return IronDoor()

    def make_fitting_expert(self):
        return Welder()


if __name__ == '__main__':

    wooden_factory = WoodenDoorFactory()
    door = wooden_factory.make_door()
    expert = wooden_factory.make_fitting_expert()
    door.get_description()
    expert.get_description()

    # I am a wooden door
    # I can only fit wooden doors

    iron_factory = IronDoorFactory()
    door = iron_factory.make_door()
    expert = iron_factory.make_fitting_expert()
    door.get_description()
    expert.get_description()

    # I am a iron door
    # I can only fit iron doors
